### ISSUE TITLE
Clang13 deprecation error

### ISSUE THREAD

**Issue by Bouncner** [gmarcais/compact_vector#6]
We have been switching our C++20 code base to clang13 and found this issue with deprecations.

I assume the problem stems from `compact_vector`?

We compile with pretty much all errors enabled + warnings as errors.

```
In file included from hyrise/third_party/compact_vector/include/compact_vector.hpp:9:
hyrise/third_party/compact_vector/include/compact_iterator.hpp:811:16: error: 'iterator<std::random_access_iterator_tag, unsigned int>' is deprecated [-Werror,-Wdeprecated-declarations]
  typedef std::iterator<std::random_access_iterator_tag, IDX> super;
               ^
hyrise/third_party/compact_vector/include/compact_vector.hpp:151:12: note: in instantiation of template class 'compact::iterator<unsigned int, 0, unsigned long long, false, 64>' requested here
  typename iterator::lhs_setter_type operator[](size_t i) {
           ^
hyrise/third_party/compact_vector/include/compact_vector.hpp:258:12: note: in instantiation of template class 'compact::vector_imp::vector<compact::vector_imp::vector_dyn<unsigned int, unsigned long long, boost::container::pmr::polymorphic_allocator<unsigned long long>, 64, false>, unsigned int, 0, unsigned long long, boost::container::pmr::polymorphic_allocator<unsigned long long>, 64, false>' requested here
  : public vector_imp::vector<vector_dyn<IDX, W, Allocator, UB, TS>, IDX, 0, W, Allocator, UB, TS>
           ^
hyrise/third_party/compact_vector/include/compact_vector.hpp:349:12: note: in instantiation of template class 'compact::vector_imp::vector_dyn<unsigned int, unsigned long long, boost::container::pmr::polymorphic_allocator<unsigned long long>, 64, false>' requested here
  : public vector_imp::vector_dyn<IDX, W, Allocator, bitsof<W>::val, false>
           ^
hyrise/src/lib/storage/vector_compression/bitpacking/bitpacking_decompressor.hpp:18:17: note: in instantiation of template class 'compact::vector<unsigned int, 0, unsigned long long, boost::container::pmr::polymorphic_allocator<unsigned long long>>' requested here
    DebugAssert(&_data == &other._data, "Cannot reassign BitPackingDecompressor");
                ^
/usr/local/opt/llvm/bin/../include/c++/v1/__iterator/iterator.h:27:29: note: 'iterator<std::random_access_iterator_tag, unsigned int>' has been explicitly marked deprecated here
struct _LIBCPP_TEMPLATE_VIS _LIBCPP_DEPRECATED_IN_CXX17 iterator
                            ^
/usr/local/opt/llvm/bin/../include/c++/v1/__config:1016:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/usr/local/opt/llvm/bin/../include/c++/v1/__config:993:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
```

**Comment by gmarcais** [comment]
I compiled the unit tests with clang13 and I do not get an error. Can you provide a smaller example of that error?

**Comment by rob-p** [comment]
Are you compiling with `-std=c++20` and `-Werror -Wdeprecated-declarations`?

**Comment by gmarcais** [comment]
Right. Forcing `-std=c++20` does give many errors. Such as:

```
ISO C++20 considers use of overloaded operator '!=' (with operand types '...' and '...')) to be ambiguous despite there being a unique best viable function with non-reversed arguments [-Werror,-Wambiguous-reversed-operator]
```

Oh C++20! Why art thou so obtuse?

Switching to C++20 is going to be more fun than I expected.

**Comment by rob-p** [comment]
Error — totally idiomatic thing now completely deprecated ... please change all code.

**Comment by gmarcais** [comment]
Well, I had avoided any `#ifdef ... #endif` shenanigans so far. No such luck moving forward.

**Comment by gmarcais** [comment]
As it turns out the deprecated issue stems from `libc++`, not `clang` on its own. The version in the develop branch compiles fine now with every C++ standards (11, 14, 17 and 20), with `g++-10` and `clang++-12`, with `libstdc++` and `libc++`.

@Bouncner, can you check that the develop branch fixes your issue?

**Comment by Bouncner** [comment]
Compiles and works like a charm! 👍

**Comment by gmarcais** [comment]
Merged as version 0.1.1
